packages/components/src/components/DiagramSequence.vue

```diff
@@ -283,6 +285,58 @@ export default {
 
       return '';
     },
+    collapseActionsForCompactLook() {
+      // Collapse any Actions that satisfy the following conditions:
+      //
+      // -Greater than 4 levels deep in the stack
+      // -All descendants are function calls (as opposed to other types
+      //  such as sql queries and external service calls)
+      // -Does not contain the Selected Action.
+
+      const isCollapsePreventingNodeType = (action: Action) =>
+        action.nodeType !== NodeType.Function && action.nodeType !== NodeType.Loop;
+
+      const isSelectedAction = (action: Action) =>
+        this.selectedEvents &&
+        this.selectedEvents.find((event) => action.eventIds?.includes(event.id));
+
+      const actionToActionSpec = new Map(
+        this.diagramSpec.actions.filter((a) => a.nodeType != 'return').map((a) => [a.action, a])
+      );
+
+      const visit = (action: Action) => {
+        const actionSpec = actionToActionSpec.get(action);
+        let descendantPreventingCollapseFound = false;
+
+        for (const child of action.children) {
+          const childHasDescendantPreventingCollapse = visit(child);
+
+          if (
+            !descendantPreventingCollapseFound &&
+            (childHasDescendantPreventingCollapse ||
+              isCollapsePreventingNodeType(child) ||
+              isSelectedAction(child))
+          )
+            descendantPreventingCollapseFound = true;
+        }
+
+        const actionShouldCollapse =
+          !descendantPreventingCollapseFound &&
+          action.children?.length > 0 &&
+          actionSpec.nodeType == 'call';
+
+        if (actionShouldCollapse) this.$set(this.collapsedActionState, actionSpec?.index, true);
+
+        return descendantPreventingCollapseFound;
+      };
+
+      const headsToVisit = this.diagramSpec.actions.filter(
+        (a) =>
+          a.ancestorIndexes.length == 4 && a.action.children?.length > 0 && a.nodeType === 'call'
```

Comment:
After messing around with this in various appmaps, I think it's best if we change this requirement to 2 instead of 4. Also, it should be `>=` instead of `==`. So:
```
 a.ancestorIndexes.length >= 2
```

Reply:
Sure.

By the way I'm grabbing heads with `a.ancestorLength == n`. All deeper nodes (`> n`, descendants of heads) are visited since I'm visiting heads recursively. I'll leave it as `==` if you agree.

Reply:
Yes, good point, you can leave it as an equality check, but can you use the strict equality operator instead?